ratelimiter/db.go

```diff
@@ -0,0 +1,126 @@
+package ratelimiter
+
+import (
+	"encoding/binary"
+	time "time"
+
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+const (
+	// BlacklistBucket is a delimiter for blacklist data.
+	BlacklistBucket byte = 1 + iota
+	// CapacityBucket is a delimiter for capacity data.
+	CapacityBucket
+)
+
+// DBInterface defines leveldb methods used by ratelimiter.
+type DBInterface interface {
+	Put(key, value []byte, wo *opt.WriteOptions) error
+	Get(key []byte, ro *opt.ReadOptions) (value []byte, err error)
+	Delete(key []byte, wo *opt.WriteOptions) error
+}
+
+// WithPrefix returns an instance of IsolatedDB.
+func WithPrefix(db DBInterface, prefix []byte) IsolatedDB {
+	return IsolatedDB{db: db, prefix: prefix}
+}
+
+// IsolatedDB adds a prefix for every operation on a database.
+type IsolatedDB struct {
+	db     DBInterface
+	prefix []byte
+}
+
+func (db IsolatedDB) withPrefix(key []byte) []byte {
+	fkey := make([]byte, len(db.prefix)+len(key))
+	copy(fkey, db.prefix)
+	copy(fkey[len(db.prefix):], key)
+	return fkey
+}
+
+// Put writes a value at the key location.
+func (db IsolatedDB) Put(key, value []byte, wo *opt.WriteOptions) error {
+	return db.db.Put(db.withPrefix(key), value, wo)
+}
+
+// Get gets a value of key.
+func (db IsolatedDB) Get(key []byte, ro *opt.ReadOptions) (value []byte, err error) {
+	return db.db.Get(db.withPrefix(key), ro)
+}
+
+// Delete record at the location of key.
+func (db IsolatedDB) Delete(key []byte, wo *opt.WriteOptions) error {
+	return db.db.Delete(db.withPrefix(key), wo)
+}
+
+// BlacklistRecord is a record with information of a deadline for a particular ID.
```

Comment:
A deadline for what? If a particular ID has a record, it means that it's blacklisted until `Deadline`?

Reply:
deadline for blacklisting a record. does it sound weird ? :)

Reply:
So it means that a particular record is blacklisted until `Deadline` (as it's a timestamp, not a duration)?

Reply:
yes